### PR TITLE
Host resolution thread pinning

### PR DESCRIPTION
**Pull request by bretambrose**
* Added resolver thread pinning via listener config
* Updated host resolution logging to make debugging easier
* Relaxed assert constraints due to suspected platform incosistencies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/host_resolver.c [hunk 1]
```diff
@@ -1020,9 +1050,21 @@ static void resolver_thread_fn(void *arg) {
                 host_entry->host_name);
             aws_mutex_unlock(&host_entry->entry_lock);
 
-            AWS_ASSERT(err_code != AWS_ERROR_SUCCESS || aws_array_list_length(&callback_address_list) > 0);
+            size_t callback_list_size = aws_array_list_length(&callback_address_list);
+            if (callback_list_size > 0) {
+                AWS_LOGF_DEBUG(
```
Comment (justinboswell) [inline source/host_resolver.c:1055]: This should probably be TRACE?

File: source/host_resolver.c [hunk 2]
```diff
@@ -942,15 +958,29 @@ static void resolver_thread_fn(void *arg) {
...
         }
+
+        if (err_code == AWS_ERROR_SUCCESS) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_IO_DNS,
+                "static, resolving host %s successful, returned %d addresses",
+                aws_string_c_str(host_entry->host_name),
+                (int)aws_array_list_length(&address_list));
+        } else {
+            AWS_LOGF_DEBUG(
```
Comment (justinboswell) [inline source/host_resolver.c:976]: Should this be a WARN/INFO?

File: source/host_resolver.c [hunk 3]
```diff
@@ -942,15 +958,29 @@ static void resolver_thread_fn(void *arg) {
...
         /* resolve and then process each record */
         int err_code = AWS_ERROR_SUCCESS;
         if (host_entry->resolution_config.impl(
                 host_entry->allocator, host_entry->host_name, &address_list, host_entry->resolution_config.impl_data)) {
 
             err_code = aws_last_error();
         }
+
+        if (err_code == AWS_ERROR_SUCCESS) {
+            AWS_LOGF_DEBUG(
```
Comment (justinboswell) [inline source/host_resolver.c:970]: Feels like TRACE?

File: source/host_resolver.c [hunk 4]
```diff
@@ -1020,9 +1050,21 @@ static void resolver_thread_fn(void *arg) {
...
-            AWS_ASSERT(err_code != AWS_ERROR_SUCCESS || aws_array_list_length(&callback_address_list) > 0);
+            size_t callback_list_size = aws_array_list_length(&callback_address_list);
+            if (callback_list_size > 0) {
+                AWS_LOGF_DEBUG(
+                    AWS_LS_IO_DNS,
+                    "static, invoking resolution callback for host %s with %d addresses",
+                    aws_string_c_str(host_entry->host_name),
+                    (int)callback_list_size);
+            } else {
+                AWS_LOGF_DEBUG(
```
Comment (justinboswell) [inline source/host_resolver.c:1061]: This log message doesn't seem accurate to what is actually happening? Isn't this just "there were no callbacks registered"?

File: include/aws/io/host_resolver.h [hunk 5]
```diff
@@ -278,6 +278,9 @@ struct aws_host_listener_options {
 
     /* User data to be passed into each callback. */
     void *user_data;
+
+    /* flags/bitfield instead? */
+    bool pin_host_entry;
```
Comment (justinboswell) [inline include/aws/io/host_resolver.h:283]: Can you document what the pinning accomplishes here?

File: source/host_resolver.c (hunk 1 above)
Comment (bretambrose) [reply source/host_resolver.c:1055]: Honestly, I don't want any host resolution logging trace.  Given the periodic per-second nature of the default host resolver, finding issues gets real ugly with trace as we've already seen with the crt http client where the customer has to use dropbox/third party service just to give us logs that cover a few minutes of light usage.

File: source/host_resolver.c (hunk 2 above)
Comment (bretambrose) [reply source/host_resolver.c:976]: We could change it to warn.

File: source/host_resolver.c (hunk 3 above)
Comment (bretambrose) [reply source/host_resolver.c:970]: Same as below

File: source/host_resolver.c (hunk 4 above)
Comment (bretambrose) [reply source/host_resolver.c:1061]: No, this is there's no addresses available

**Review by rccarper — APPROVED** [review]
Approved

**Pull request merged**